non-project mode deprecation blog (#3952)
pulumi/pulumi-hugo: themes/default/content/blog/deprecating-non-project-mode/index.md

---
title: "Deprecating non-project mode from DIY backends"

> date: 2024-02-19

Reviewer: May want to bump the date when you're ready to ship!

> meta_desc: Read about the plans for the removal of non-project mode.

Reviewer: Just a note to remember to include a meta image.

```suggestion
meta_desc: Read about the plans for the removal of non-project mode.
meta_image: meta.png
```

authors:
    - fraser-waters
tags:
    - features
    - pulumi-releases
---

> Last year we released [project support for self-managed backends](/blog/project-scoped-stacks-in-self-managed-backend/), we now plan to remove support for non-project mode.

Reviewer: ```suggestion
Last year we released [project support for self-managed backends](/blog/project-scoped-stacks-in-self-managed-backend/). We now plan to remove support for non-project mode.
```

> Last year we released [project support for self-managed backends](/blog/project-scoped-stacks-in-self-managed-backend/), we now plan to remove support for non-project mode.

Reviewer 2: It feels like more text would be helpful here. Why are we now deprecating non-project mode? While the linked blog explains things, it feels like we should take a moment and help provide some context/background for those people who haven't read (or don't wish to read) the other blog post. Right now this first paragraph is too abrupt (IMO).

Author: Added a bit more. This is a systems simplification, we don't think there's any use cases for non-project mode and that project mode is better for everyone. Removing it like this lets us push any stragglers towards it to have a better experience, and makes the system less costly to understand (the backends are much more consistent, don't need to know about non-project mode and the oddities it entails), and less costly to maintain (we can remove a load of code that really no one needs anymore).

<!--more-->

> Users using a DIY backend (previously referred to as self-managed) may have seen a warning in recent releases of Pulumi:

Reviewer: Just wondering -- have we renamed these from "self-managed backends" to "DIY backends"?

Author: Yes

```
warning: Legacy non-project mode support for DIY backends are due to be deprecated this year.
Please migrate to project mode by running 'pulumi state upgrade'.
You can disable this warning by setting PULUMI_DIY_BACKEND_ACKNOWLEDGE_LEGACY_WARNING to true.
```

> To assuage any concerns about this warning it is _specifically_ about running the DIY backend in non-project mode where all the state files are kept in one top level folder, we are _not_ removing support for DIY backends in general.

Reviewer: ```suggestion
To assuage any concerns about this warning, it is _specifically_ about running the DIY backend in non-project mode, where all the state files are kept in one top level folder. We are _not_ removing support for DIY backends in general.
```

As described in the warning we plan on deprecating and removing support for non-project mode this year.

## Phase 1 - Warning

> The first phase of this deprecation has already been released. When opening a DIY state store if the system see it is still in non-project mode the above warning is printed. You can set the environment variable `PULUMI_DIY_BACKEND_ACKNOWLEDGE_LEGACY_WARNING` to silence this warning.

Reviewer: ```suggestion
The first phase of this deprecation has already been released. When opening a DIY state store, if the system sees it is still in non-project mode, the above warning is printed. You can set the environment variable `PULUMI_DIY_BACKEND_ACKNOWLEDGE_LEGACY_WARNING` to silence this warning.
```

## Phase 2 - Soft error

> In a few months we plan to make this a soft error. Opening a state store in non-project mode will result in an error like:

Reviewer 3: ```suggestion
In a few months we plan to make this a soft error. Opening a state store in non-project mode will result in an error like:
```

Would it make sense to make this a little bit more concrete, e.g. "In 2-3 months", so users have a more concrete idea of the changes that are coming?  A few months feels a little bit vague for users planning purposes.  Similar for the other steps below.

```
error: Legacy non-project mode support for DIY backends are due to be deprecated this year.
Please migrate to project mode by running 'pulumi state upgrade'.
You can disable this error by setting PULUMI_DIY_BACKEND_ACKNOWLEDGE_LEGACY_ERROR to true.
```

> Note that this is an error. Without the new environment variable set this will cause CLI execution to stop. But this is only a soft error, users will be able to set the new environment variable `PULUMI_DIY_BACKEND_ACKNOWLEDGE_LEGACY_ERROR` to make the CLI skip this error.

Reviewer: >this will cause CLI execution to stop

May be worth noting (assuming this is the case) that execution will stop before performing any updates.

## Phase 3 - Hard error

> 2-3 months after that this will become a hard error.

Reviewer: ```suggestion
In 2-3 months after Phase 2, this will become a hard error.
```

```
error: Legacy non-project mode support for DIY backends are deprecated.
Please migrate to project mode by running 'pulumi state upgrade'.
```

There will be no way to skip this error. Users on this version of the CLI _must_ upgrade their state to continue using Pulumi.

## Phase 4 - Removal of feature

> 2-3 months after that all support for non-project mode will be removed. This will include the capability in the CLI to do a state upgrade from non-project mode.

Reviewer: ```suggestion
In 2-3 months after Phase 3, all support for non-project mode will be removed. This will include the capability in the CLI to do a state upgrade from non-project mode.
```

```
error: Legacy non-project mode support for DIY backends are unsupported.
```

> If users encounter this error they'll have to download an older version of the CLI to do the state upgrade. We're hoping that no one gets to this point given the numerous warnings before hand.

Reviewer: Since some may Google their way to this post, providing a link to the download page would help -- although how will they know which version to downgrade to? 

```suggestion
If users encounter this error, they'll have to [downgrade to an older version of the CLI](/docs/install/versions/) to do the state upgrade. We're hoping that no one gets to this point given the numerous warnings beforehand.
```
